packages/contracts/src/contracts/current/forwarder/Forwarder.sol

```diff
@@ -0,0 +1,315 @@
+    /// @param amount The amount of ZRX to withdraw.
+    function withdraw(uint amount) public {
+        require(balanceOf[msg.sender] >= amount);
+        balanceOf[msg.sender] -= amount;
+        ZRX_TOKEN.transfer(msg.sender, amount);
+    }
+
+    /// @dev Buys the fee tokens as well as any fees required to buy the requested fee tokens.
+    ///      It is possible that a request to buy 200 ZRX fee tokens will require purchasing 202 ZRX tokens
+    ///      As 2 ZRX is required to purchase the 200 ZRX fee tokens.
+    /// @param orders An array of Order struct containing order specifications for fees.
+    /// @param signatures An array of Proof that order has been created by maker for the fee orders.
+    /// @param feeAmount The number of requested ZRX fee tokens.
+    /// @return Amounts filled and fees paid by maker and taker. makerTokenAmount is the zrx amount deducted of fees
+    function buyFeeTokens(
+        Order[] memory orders,
+        bytes[] memory signatures,
+        uint256 feeAmount)
+        private
+        returns (Exchange.FillResults memory totalFillResult)
+    {
+        address token = readAddress(orders[0].makerAssetData, 1);
+        require(token == address(ZRX_TOKEN), ERROR_INVALID_INPUT);
+        for (uint256 i = 0; i < orders.length; i++) {
+            // Token being bought by taker must be the same for each order
+            require(areBytesEqual(orders[i].makerAssetData, orders[0].makerAssetData));
+
+            // Calculate the remaining amount of makerToken to buy
+            uint256 remainingMakerTokenFillAmount = safeSub(feeAmount, totalFillResult.makerAssetFilledAmount);
+
+            // Convert the remaining amount of makerToken to buy into remaining amount
+            // of takerToken to sell, assuming entire amount can be sold in the current order
+            uint256 remainingTakerTokenFillAmount = getPartialAmount(
+                orders[i].takerAssetAmount,
+                safeSub(orders[i].makerAssetAmount, orders[i].takerFee), // our exchange rate after fees 
+                remainingMakerTokenFillAmount);
+
+            // Attempt to sell the remaining amount of takerToken
+            // Round up the amount to ensure we don't under buy by a fractional amount
+            Exchange.FillResults memory singleFillResult = EXCHANGE.fillOrder(
+                orders[i],
+                safeAdd(remainingTakerTokenFillAmount, 1),
+                signatures[i]
+            );
+
+            // We didn't buy the full amount when buying ZRX as some were taken for fees
+            singleFillResult.makerAssetFilledAmount = safeSub(singleFillResult.makerAssetFilledAmount, singleFillResult.takerFeePaid);
+            // Update amounts filled and fees paid by maker and taker
+            addFillResults(totalFillResult, singleFillResult);
+
+            // Stop execution if the entire amount of makerToken has been bought
+            if (totalFillResult.makerAssetFilledAmount == feeAmount) {
```

Comment:
Couldn't `totalFillResult.makerAssetFilledAmount` > `feeAmount`, since we are over-buying?